File: Week 6 - Python/sentimental-credit/credit.py
from cs50 import get_string


def main():
    num = get_string("Number: ")
    if check_valid(num):
        print_brand(num)
    else:
        print("INVALID")


def check_valid(n):
    if len(n) in [13, 15, 16] and check_sum(n):
        return True
    return False


def check_sum(n):
    sum = 0
    for i, c in enumerate(reversed(n)):
        if i % 2 == 0:
            sum += int(c)
        else:
            for j in str(int(c) * 2):
                sum += int(j)

    if sum % 10 == 0:
        return True
    return False


def print_brand(n):
    if len(n) == 15 and n[:2] in ["34", "37"]:
        print("AMEX")
    elif len(n) == 16 and n[:2] in ["51", "52", "53", "54", "55"]:
        print("MASTERCARD")
    elif (len(n) == 13 or len(n) == 16) and n[0] == '4':
        print("VISA")
    else:
        print("INVALID")


main()

File: Week 1 - C/credit/credit.c
#include <cs50.h>
#include <stdio.h>

int find_length(long long n);
bool checksum(long long ccn);
bool check_validity(long long card_number);
void print_credit_card_brand(long long ccn);

int main(void)
{
    long long card_number;
    do
    {
        card_number = get_long_long("Number: ");
    }
    while (card_number <= 0);

    if (check_validity(card_number))
        print_credit_card_brand(card_number);
    else
        printf("INVALID\n");
}

bool check_validity(long long card_number)
{
    int len = find_length(card_number);

    return (len == 13 || len == 15 || len == 16) && checksum(card_number);
}

int find_length(long long n)
{
    int len;
    for (len = 0; n != 0; n /= 10, len++)
        ;
    return len;
}

bool checksum(long long ccn)
{
    int sum = 0;
    for (int i = 0; ccn != 0; i++, ccn /= 10)
    {
        if (i % 2 == 0)
            sum += ccn % 10;
        else
        {
            int digit = 2 * (ccn % 10);
            sum += digit / 10 + digit % 10;
        }
    }
    return (sum % 10) == 0;
}

void print_credit_card_brand(long long ccn)
{
    if ((ccn >= 34e13 && ccn <= 35e13) || (ccn >= 37e13 && ccn < 38e13))
        printf("AMEX\n");

    else if (ccn >= 51e14 && ccn < 56e14)
        printf("MASTERCARD\n");

    else if ((ccn >= 4e12 && ccn < 5e12) || (ccn >= 4e15 && ccn < 5e15))
        printf("VISA\n");

    else
        printf("INVALID\n");
}
